### PR TITLE
[WIP] new wire protocol

### PR DESCRIPTION
**Pull request by majestrate**
don't merge me yet

### CODE REVIEW

File: docs/wire-protocol.txt [hunk 1]
```diff
@@ -257,7 +284,7 @@ request protocol upgrade
 <1 byte protocol max version to upgrade to>
 <remaining bytes discarded>
 
-command: 'V' (version upgrade)
+command: 'v' (version upgrade)
```
Comment (michael-loki) [inline docs/wire-protocol.txt]: Is there a reason to change all the casings here? Is it to help migration?

File: llarp/link/iwp.cpp [hunk 2]
```diff
@@ -1,25 +1,165 @@
...
-      return nullptr;
+      using namespace std::placeholders;
+      return NewServer(
+          r->crypto(), r->encryption(), std::bind(&AbstractRouter::rc, r),
+          std::bind(&AbstractRouter::HandleRecvLinkMessageBuffer, r, _1, _2),
+          std::bind(&AbstractRouter::OnSessionEstablished, r, _1),
+          std::bind(&AbstractRouter::CheckRenegotiateValid, r, _1, _2),
+          std::bind(&AbstractRouter::Sign, r, _1, _2),
+          std::bind(&AbstractRouter::OnConnectTimeout, r, _1),
+          std::bind(&AbstractRouter::SessionClosed, r, _1));
```
Comment (michael-loki) [inline llarp/link/iwp.cpp]: this is making me feel like `AbstractRouter` is a bit too long. `IRouter` might have been better?

File: llarp/link/iwp.cpp [hunk 3]
```diff
@@ -1,25 +1,165 @@
...
-              llarp::SessionRenegotiateHandler, llarp::SignBufferFunc,
-              llarp::TimeoutHandler, llarp::SessionClosedHandler)
+    NewServer(Crypto* c, const SecretKey& enckey, GetRCFunc getrc,
+              LinkMessageHandler h, SessionEstablishedHandler est,
+              SessionRenegotiateHandler reneg, SignBufferFunc sign,
+              TimeoutHandler t, SessionClosedHandler closed)
     {
-      // TODO: implement me
-      return nullptr;
+      return std::unique_ptr< ILinkLayer >(
```
Comment (michael-loki) [inline llarp/link/iwp.cpp]: FYI `return std::make_unique<LinkLayer>(c, .....);` is a safer option in the general case (doesn't make a difference here)

File: llarp/link/iwp_internal.hpp [hunk 4]
```diff
@@ -17,6 +18,117 @@ namespace llarp
   {
     struct LinkLayer;
 
+    using FlowID_t = llarp::AlignedBuffer< 32 >;
```
Comment (michael-loki) [inline llarp/link/iwp_internal.hpp]: I would do `struct FlowID_t : public AlignedBuffer< 32 > {};` here, it helps prevent implicit conversions between other typedefs to the same size `AlignedBuffer`.

File: docs/wire-protocol.txt (hunk 1 above)
Comment (majestrate) [reply docs/wire-protocol.txt]: disctinction from inner and outer message commands, stylistic shtick.

File: llarp/link/iwp.cpp (hunk 2 above)
Comment (majestrate) [reply llarp/link/iwp.cpp]: probably better `IRouter` but it's too late :DDDDDDDDD

File: llarp/link/iwp.cpp (hunk 3 above)
Comment (majestrate) [reply llarp/link/iwp.cpp]: this code is not done yet.

File: llarp/link/iwp_internal.hpp (hunk 4 above)
Comment (majestrate) [reply llarp/link/iwp_internal.hpp]: it's actually supposed to be a 16 byte thing according to spec so this change is out dated.

**Comment by majestrate** [comment]
migration path is we enable a separate transport of different rank and then switch the rank when it's time to use it as primary.

**Comment by neuroscr** [comment]
We should not run two protocols in parallel. Just hard cut over.

**Pull request merged**